obspy/signal/cross_correlation.py

```diff
@@ -151,22 +187,128 @@ def correlate(a, b, shift, demean=True, normalize=True, domain='freq'):
+
+    :type data: :class:`~numpy.ndarray`, :class:`~obspy.core.trace.Trace`
+    :param data: first signal
+    :type template: :class:`~numpy.ndarray`, :class:`~obspy.core.trace.Trace`
+    :param template: second signal to correlate with first signal.
+        Its length must be smaller or equal to the length of ``data``.
+    :param str mode: correlation mode to use.
+        It is passed to the used correlation function.
+        See :func:`scipy.signal.correlate` for possible options.
+        The parameter determines the length of the correlation function.
+    :param normalize:
+        One of ``'naive'``, ``'full'`` or ``None``.
+        ``'full'`` normalizes every correlation properly,
+        whereas ``'naive'`` normalizes by the overall standard deviations.
+        ``None`` does not normalize.
+    :param demean: Demean data beforehand. For ``normalize='full'`` data is
+        demeaned in different windows for each correlation value.
+    :param str method: Method to use to calculate the correlation.
+         ``'direct'``: The correlation is determined directly from sums,
+         the definition of correlation.
+         ``'fft'`` The Fast Fourier Transform is used to perform the
+         correlation more quickly.
+         ``'auto'`` Automatically chooses direct or Fourier method based on an
+         estimate of which is faster. (Only availlable for SciPy versions >=
+         0.19. For older Scipy version method defaults to ``'fft'``.)
+
+    :return: cross-correlation function.
+
+    .. note::
+        Calling the function with ``demean=True, normalize='full'`` (default)
+        returns the zero-normalized cross-correlation function.
+        Calling the function with ``demean=False, normalize='full'``
+        returns the normalized cross-correlation function.
+
+    .. rubric:: Example
+
+    >>> from obspy import read
+    >>> data = read()[0]
+    >>> template = data[450:550]
+    >>> cc = correlate_template(data, template)
+    >>> index = np.argmax(cc)
+    >>> index
+    450
+    >>> round(cc[index], 9)
+    1.0
+    """
+    # if we get Trace objects, use their data arrays
+    if isinstance(data, Trace):
+        data = data.data
+    if isinstance(template, Trace):
+        template = template.data
+    data = np.asarray(data)
+    template = np.asarray(template)
+    lent = len(template)
+    if len(data) < lent:
+        raise ValueError('Data must not be shorter than template.')
+    if demean:
+        template = template - np.mean(template)
+        if normalize != 'full':
+            data = data - np.mean(data)
+    cc = _call_scipy_correlate(data, template, mode, method)
+    if normalize is None:
+        norm = 1
     else:
-        raise ValueError("domain keyword has to be one of ('freq', 'time')")
-    return _xcorr(a, b, shift, domain=domain) / stdev
+        norm = np.sum(template ** 2)
+        if normalize == 'naive':
+            norm = (norm * np.sum(data ** 2)) ** 0.5
```

Comment:
Some things here could happen in place to make it faster at the cost of more ugly code. I'm not sure this is a worthy trade-off here but it would be possible. For many things an optional `numexpr` dependency would also be really useful but I'm not sure we want to go down this road.

Reply:
I changed it to in-place operations with the original lines as comments.